### ISSUE TITLE
[AArch64] Redundant masks for umull and umaddl

### ISSUE THREAD

**Issue by nikic** [llvm/llvm-project#98481]
See https://llvm.godbolt.org/z/Gbe8xPzd9:
```llvm
define i64 @umull(i64 %x) {
    %lo = and i64 %x, u0xffffffff
    %hi = lshr i64 %x, 32
    %mul = mul i64 %lo, %hi
    ret i64 %mul
}

define i64 @umull_no_mask(i64 %x, i64 %y) {
    %lo = and i64 %x, u0xffffffff
    %hi = and i64 %y, u0xffffffff
    %mul = mul i64 %lo, %hi
    ret i64 %mul
}

define i64 @umaddl(i64 %x, i64 %a) {
    %lo = and i64 %x, u0xffffffff
    %hi = lshr i64 %x, 32
    %mul = mul i64 %lo, %hi
    %add = add i64 %a, %mul
    ret i64 %add
}

define i64 @umaddl2(i64 %x, i64 %y, i64 %a) {
    %lo = and i64 %x, u0xffffffff
    %hi = and i64 %y, u0xffffffff
    %mul = mul i64 %lo, %hi
    %add = add i64 %a, %mul
    ret i64 %add
}
```

Produces:
```asm
umull:                                  // @umull
        lsr     x8, x0, #32
        and     x9, x0, #0xffffffff
        umull   x0, w9, w8
        ret
.Lfunc_end0:
umull_no_mask:                          // @umull_no_mask
        umull   x0, w0, w1
        ret
.Lfunc_end1:
umaddl:                                 // @umaddl
        lsr     x8, x0, #32
        and     x9, x0, #0xffffffff
        umaddl  x0, w9, w8, x1
        ret
.Lfunc_end2:
umaddl2:                                // @umaddl2
        and     x8, x0, #0xffffffff
        and     x9, x1, #0xffffffff
        umaddl  x0, w8, w9, x2
        ret
.Lfunc_end3:
```

The masking is redundant in all cases. Note that it does get omitted for umull if *both* operands are mask, but not if only one is. And for umaddl it does not happen even if both are masked.

**Comment by davemgreen** [comment]
The tablegen patterns we added are based on KnownBits, so don't necessarily remove the redundant instructions. https://github.com/llvm/llvm-project/pull/96293 was to capture cases that were cross-basic-block but helped for the signed cases. It could presumably be extended to `And 0xffffffff` too, if we didn't try to fix the tablegen patterns instead.

**Comment by shepmaster** [comment]
Adding a little bit of background, I found this when porting xxHash from C to Rust and noticed that I was always missing a good 5% of performance. Digging into the C implementation, I found an [educational note](https://github.com/Cyan4973/xxHash/blob/d5fe4f54c47bc8b8e76c6da9146c32d5c720cd79/xxhash.h#L5595-L5610) that pointed to the core problem. In the Rust implementation, avoiding these unneeded `and` operations raises the throughput from 19 GiB/sec to 20GiB/sec on my M1 Pro.
